cmd/gossamer/prune.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	base := state.NewBaseState(db)
+
+	bestHash, err := base.LoadBestBlockHash()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get best block hash: %w", err)
+	}
+
+	// load blocktree
+	bt := blocktree.NewEmptyBlockTree(db)
+	if err = bt.Load(); err != nil {
+		return nil, fmt.Errorf("failed to load blocktree: %w", err)
+	}
+
+	// create blockState state
+	blockState, err := state.NewBlockState(db, bt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create block state: %w", err)
+	}
+
+	// create bloom filter
+	bloom, err := newStateBloomWithSize(bloomSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new bloom filter of size %d %w", bloomSize, err)
+	}
+
+	// load storage state
+	storageState, err := state.NewStorageState(db, blockState, trie.NewEmptyTrie())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new storage state %w", err)
+	}
+
+	return &Pruner{
+		inputDB:        db,
+		storageState:   storageState,
+		blockState:     blockState,
+		bloom:          bloom,
+		bestBlockHash:  bestHash,
+		retainBlockNum: retainBlockNum,
+	}, nil
+}
+
+func (p *Pruner) setBloomFilter() error {
+	// latest block header
+	header, err := p.blockState.GetHeader(p.bestBlockHash)
+	if err != nil {
+		return err
+	}
+
+	latestBlockNum := header.Number.Int64()
+	keys := make(map[common.Hash]interface{})
+
+	logger.Info("Latest block number", "num", latestBlockNum)
+
+	if latestBlockNum-p.retainBlockNum <= 0 {
+		return fmt.Errorf("not enough block to perform pruning")
+	}
+
+	// loop from latest to last 256
+	for blockNum := header.Number.Int64(); blockNum > 0 && blockNum >= latestBlockNum-p.retainBlockNum; {
+		tr, err := p.storageState.LoadFromDB(header.StateRoot)
+		if err != nil {
+			return err
+		}
+
+		err = tr.GetDBKey(tr.RootNode(), keys)
+		if err != nil {
+			return err
+		}
+
+		// get parent header of current block
+		header, err = p.blockState.GetHeader(header.ParentHash)
+		if err != nil {
+			return err
+		}
+		blockNum = header.Number.Int64()
+	}
+
+	for key := range keys {
+		err = p.bloom.put(key.ToBytes())
+		if err != nil {
+			return err
+		}
+	}
+
+	fmt.Println("Total keys added in bloom filter", len(keys))
+
+	logger.Info("Total keys added in bloom filter", "keysCount", len(keys))
+	return nil
+}
+
+func (p *Pruner) prune(inDBPath, pruneDBPath string) error {
+	var err error
+	p.prunedDB, err = loadBadgerDB(inDBPath)
+	if err != nil {
+		return fmt.Errorf("failed to load badger DB %w", err)
+	}
+
+	defer p.prunedDB.Close()
+
+	if err = p.streamDB(pruneDBPath); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Pruner) streamDB(outDir string) error {
+	outOptions := badger.DefaultOptions(outDir)
+	outOptions.WithInMemory(false)
+	outOptions.Dir = outDir
+
+	// Open output DB.
+	outDB, err := badger.Open(outOptions)
+	if err != nil {
+		return fmt.Errorf("cannot open out DB at %s error %w", outDir, err)
+	}
+
+	defer outDB.Close()
+
+	writer := outDB.NewStreamWriter()
```

Comment:
what does the stream writer do? is this code writing the database contents to a new database?

Reply:
Yes. It will stream all data except the pruned keys to the new DB. In case the pruning goes wrong user would still have the actual DB.
Logic for selecting keys is written in 
```go
stream.ChooseKey = func(item *badger.Item) bool {
		key := string(item.Key())
		// All the non storage keys will be streamed to new db.
		if !strings.HasPrefix(key, StoragePrefix) {
			return true
		}

		// Only keys present in bloom filter will be streamed to new db
		key = strings.TrimPrefix(key, StoragePrefix)
		exist := p.bloom.contain([]byte(key))
		return exist
	}
```